### ISSUE TITLE
[QUESTION] nginx reverse proxy setup

### ISSUE THREAD

**Issue by ras07** [benbusby/whoogle-search#678]
I'm running the Whoogle Docker container behind an nginx HTTPS proxy (in another container). The way I have it set up, nginx listens on port 443 but nothing's listening on 80 (I'd prefer not to even redirect 80 to 443, so that there's no way to send an unencrypted search request to the server). Note, there's several moving parts here, but I don't _think_ the problem has anything to do with Docker; I'm pretty sure it's between Whoogle and nginx.

The problem is that the first time I access https://whoogle.mydomain.com after not using it for a while (10 minutes or so) I get an error; but if I then retry the exact same request it works fine. Then it keeps working until I don't use it for a while.

With Brave (or Chrome) the error says:
```
This page isn’t working
whoogle.mydomain.com didn’t send any data.
ERR_EMPTY_RESPONSE
```
In the URL bar is a link that looks something like `http://whoogle.mydomain.com/session/941d51d5-9af7-4ac8-4f21-374a0cb2c024?follow=https%3A%2F%2Fwhoogle.mydomain.com%2F` Note that it's now http, not https. (But the follow= is https.) Hitting the "Reload" button doesn't work (unless I manually add https:// to the front of the link first), but retrying the base https URL does.

I don't quite understand why it all starts working the second time. If I open Whoogle in a private browser window it never works the first time (but again, works fine after that). If I close the private browser window and immediately open another one, it always fails (and then works fine on the second and subsequent tries from that window).

My nginx config looks like:
```server {
	listen		443	ssl http2;
	server_name	whoogle.mydomain.com;
	location / {
		limit_except		GET HEAD POST PUT	{ deny all; }
		proxy_set_header	Host			$host;
		proxy_set_header	X-Forwarded-For		$proxy_add_x_forwarded_for;
		proxy_set_header	X-Forwarded-Proto	$scheme;
		proxy_read_timeout	300;
		proxy_connect_timeout	300;
		proxy_send_timeout	300;
		proxy_pass	http://dkr-whoogle:5000;    # dkr-whoogle is the name of my docker container; they share a docker network
	}
}
```
Here's the key thing - if I listen on port 80 and just redirect it to 443, like so in the nginx config:
```
server {
	listen			80;
	server_name		whoogle.mydomain.com;
	return 301		https://$host$request_uri;
}
```
Then everything starts working fine. But again, I'd rather not do that; it lends itself to sending search text in the clear.

I don't see anything meaningful in the Whoogle log. If I look in the nginx access log, on the first (failed) access I see a single line that says `"GET / HTTP/2.0" 307 401 "-" "Mozilla/5.0 (etc)`. 307 is "Temporary redirect" - seems like that is the root of the problem? I tried to follow the request in the Developer Tools of Brave (Chrome); I think what's going on is: 

- My browser makes the request to the proxy via https
- The proxy makes an http request to Whoogle
- Whoogle returns an http redirect that has the session token
- The proxy server sends on the http redirect back to the browser, unchanged
- My browser follows the redirect by making an http request, but nothing's listening on http.

Anybody know of a way to make nginx handle this? If the HTTP response code was 302 instead of 307, I think it might just handle it transparently (not sure about that though). In Apache I think you could handle it with mod_proxy and mod_rewrite, but I don't know if this is possible in nginx.

**Comment by 0xspade** [comment]
Hi,

Here's my nginx config `/etc/nginx/sites-enabled/whoogle`:
```
server {
    server_name  whoogle.your.domain;
    if ($host != "whoogle.your.domain") {
        return 301 https://www.youtube.com/watch?v=dQw4w9WgXcQ;
    }
    location / {
        proxy_pass http://localhost:1337;
        proxy_set_header Host $host;
        add_header X-XSS-Protection "0" always;
        add_header Cache-Control 'no-store, no-cache, must-revalidate, proxy-revalidate, max-age=0';
        add_header X-Frame-Options DENY always;
        add_header X-Content-Type-Options "nosniff";
        add_header Referrer-Policy "same-origin";
    }
}
```
Restart nginx and you should be good to go.

For the docker CLI command, here's mine. 
`docker run -d -p 1337:1337 -e WHOOGLE_AUTOCOMPLETE=0 -e WHOOGLE_ALT_WIKI="wikiless.org" -e WHOOGLE_ALT_IMG="imgin.voidnet.tech" -e WHOOGLE_ALT_TL="lingva.ml" -e WHOOGLE_ALT_MD="scribe.bus-hit.me" -e WHOOGLE_CONFIG_LANGUAGE="lang_en" -e WHOOGLE_CONFIG_SEARCH_LANGUAGE="lang_en" -e WHOOGLE_CONFIG_NEW_TAB=1 -e WHOOGLE_CONFIG_DISABLE=1 -e WHOOGLE_CONFIG_ALTS=1 -e WHOOGLE_ALT_TW="nitter.snopyta.org" -e WHOOGLE_ALT_YT="invidious.snopyta.org" -e WHOOGLE_ALT_IG="bibliogram.art/u" -e WHOOGLE_ALT_RD="libredd.it" -e WHOOGLE_CONFIG_STYLE=":root { --whoogle-logo:#685e79;--whoogle-page-bg:#ffffff;--whoogle-element-bg:#4285f4;--whoogle-text:#000000;--whoogle-contrast-text:#ffffff;--whoogle-secondary-text:#70757a;--whoogle-result-bg:#ffffff;--whoogle-result-title:#1967d2;--whoogle-result-url:#0d652d;--whoogle-result-visited:#4b11a8;--whoogle-dark-logo:#ffffff;--whoogle-dark-page-bg:#111111;--whoogle-dark-element-bg:#303134;--whoogle-dark-text:#dddddd;--whoogle-dark-contrast-text:#e8eaed;--whoogle-dark-secondary-text:#bdc1c6;--whoogle-dark-result-bg:#222222;--whoogle-dark-result-title:#8ab4f8;--whoogle-dark-result-url:#34a853;--whoogle-dark-result-visited:#c58af9;}#whoogle-w{fill:#4285f4;}#whoogle-h{fill:#ea4335;}#whoogle-o-1{fill:#fbbc05;}#whoogle-o-2{fill:#4285f4;}#whoogle-g{fill:#34a853;}#whoogle-l{fill:#ea4335;}#whoogle-e{fill:#fbbc05;}" -e EXPOSE_PORT=1337 --restart unless-stopped --name whoogle benbusby/whoogle-search:latest`

Believe it or not, even though my docker CLI command is very long, it still works like a charm.

LMK if this works.

**Comment by 0xspade** [comment]
As a side note, I use Cloudflare to manage my whoogle stuff, including SSL and the traffic just in case my whoogle instance gets DDoSed.

**Comment by Albonycal** [comment]
My network was DDOSed twice,
[image: image.png]
why do they want to DDOS an archlinux mirror, lol

**Comment by 0xspade** [comment]
@Albonycal use CF as your DNS, it's free.

**Comment by Albonycal** [comment]
already started using that, but still some stuff got through.
Fixed that tho

**Comment by ras07** [comment]
> Hi,
> 
> Here's my nginx config `/etc/nginx/sites-enabled/whoogle`:
> 
> ```
> server {
>     server_name  whoogle.your.domain;
>     if ($host != "whoogle.your.domain") {
>         return 301 https://www.youtube.com/watch?v=dQw4w9WgXcQ;
>     }
>     location / {
>         proxy_pass http://localhost:1337;
>         proxy_set_header Host $host;
>         add_header X-XSS-Protection "0" always;
>         add_header Cache-Control 'no-store, no-cache, must-revalidate, proxy-revalidate, max-age=0';
>         add_header X-Frame-Options DENY always;
>         add_header X-Content-Type-Options "nosniff";
>         add_header Referrer-Policy "same-origin";
>     }
> }
> ```

What ports are  you listening on?

I'm trying to avoid having anything at all that's HTTP.

**Comment by ras07** [comment]
SOLVED: I did eventually find a workable solution. As I mentioned in the OP, the sequence of events was;

> * My browser makes the request to the proxy via https
> * The proxy makes an http request to Whoogle
> * Whoogle returns an http redirect that has the session token
> * The proxy server sends on the http redirect back to the browser, unchanged
> * My browser follows the redirect by making an http request, but nothing's listening on http.

I wanted to find a way to change that second-to-last step; in other words, have the proxy "fix" the redirect to say **https://** rather than **http://**.

I tried to do it with the sub_filter command, but for reasons I still don't understand I couldn't make it work. However what did work is:
```
proxy_set_header 	Accept-Encoding "";
proxy_redirect		http://$host https://$host;
```
in the server block. (tbh I'm not sure the **proxy_set_header** command is necessary.)

I should note that without doing this, when you set Whoogle as your primary search engine and submit a query from the URL entry box, the query parameters end up going out via unencrypted HTTP, _even if you the root URL and the browser settings are set to use HTTPS._
